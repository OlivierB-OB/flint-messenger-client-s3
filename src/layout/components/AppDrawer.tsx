import Box from '@material-ui/core/Box';
import Drawer from '@material-ui/core/Drawer';
import IconButton from '@material-ui/core/IconButton';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import ArrowBackIos from '@material-ui/icons/ArrowBackIos';
import React from 'react';
import { connect } from 'react-redux';
import { Dispatch, Action } from 'redux';
import { IAppState } from '../../appReducer';
import { IDrawerContent } from '../types';
import { hideDrawer } from '../actions/hideDrawer';
import { MyContacts } from '../../users/components/MyContacts';
import { MyConversations } from '../../conversations/components/MyConversations';

export interface IDrawerDisplayProps {
  show: boolean;
  content?: IDrawerContent;
  hideDrawer: () => void;
}

export const drawerWidth = 500;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    drawerHeader: {
      textAlign: 'right',
      position: 'sticky',
      top: 0,
      zIndex: 100,
      backgroundColor: theme.palette.background.paper,
      borderBottom: `1px solid ${theme.palette.divider}`,
    },
    paper: {
      width: drawerWidth,
    },
  }),
);

export function DrawerDisplay({ show, content, hideDrawer }: IDrawerDisplayProps) {
  const { drawerHeader, paper } = useStyles();
  const contentDisplay =
    content === 'contacts' ? <MyContacts /> : content === 'conversations' ? <MyConversations /> : null;
  return (
    <Drawer variant="persistent" anchor="left" open={show} onClose={hideDrawer} classes={{ paper }}>
      <Box className={drawerHeader}>
        <IconButton aria-label="collapse" onClick={hideDrawer}>
          <ArrowBackIos />
        </IconButton>
      </Box>
      <Box>{contentDisplay}</Box>
    </Drawer>
  );
}

const mapStateToProps = ({ layout }: IAppState) => ({
  show: layout.showDrawer,
  content: layout.drawerContent,
});

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  hideDrawer: () => dispatch(hideDrawer()),
});

export const AppDrawer = connect(mapStateToProps, mapDispatchToProps)(DrawerDisplay);
