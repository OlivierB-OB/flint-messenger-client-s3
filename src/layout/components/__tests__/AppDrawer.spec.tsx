import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { AppDrawer } from '../AppDrawer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('AppDrawer', () => {
  let Connected: any = AppDrawer;

  it('should provide show content, and a closing callback to DrawerDisplay', async () => {
    const store = mockStore({
      layout: {
        show: 'show from state',
        content: 'content from state',
      },
    });
    const component = shallow(<Connected store={store} />).dive();
    expect(component).toMatchSnapshot();
    (component.prop('hideDrawer') as any)();
    expect(store.getActions()).toEqual([
      {
        showDrawer: false,
        type: 'TOGGLE_DRAWER',
      },
    ]);
    store.clearActions();
  });
});
