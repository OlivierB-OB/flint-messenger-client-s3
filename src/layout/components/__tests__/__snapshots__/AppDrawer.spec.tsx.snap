// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppDrawer should provide show content, and a closing callback to DrawerDisplay 1`] = `
<DrawerDisplay
  hideDrawer={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
